File: Contribute/Plugin Samples/Plugin Samples/C#/Parser/XMLParserPlugin/3rd Party Control/XmlTreeNode.cs
using System;
using System.Xml;
using System.Windows.Forms;
using SS.Ynote.Engine.Framework;

namespace XMLTreeView
{
	/// <summary>
	/// Summary description for XmlTreeNode.
	/// </summary>
	internal class XmlTreeNode : TreeNode
	{
		#region Fields
		private XmlNode elem = null;
		private string childrenXml = string.Empty;
		private string indent = string.Empty;
		private bool hitEnd = false;
		private bool hitStart = false;
		#endregion // Fields

		#region Constructors
		internal XmlTreeNode( string text, XmlNode elem ) : base( text )
		{
			this.elem = elem;
		}
		#endregion // Constructors

		#region Properties
		internal XmlNode ConnectedXmlElement
		{
			get
			{
				return elem;
			}
		}

		internal string SelfAndChildren
		{
			get
			{
				return childrenXml;
			}
		}

		public new string Text
		{
			get
			{
				return base.Text;
			}
			set
			{
				if ( value != null && value.Length > 0 )
				{
					try
					{
						if ( value != childrenXml )
						{
							XmlDocumentFragment frag = elem.OwnerDocument.CreateDocumentFragment();
							frag.InnerXml = value;
							elem.ParentNode.ReplaceChild( frag, elem );
							InnerXmlTreeView innerView = (InnerXmlTreeView) TreeView;
							innerView.Xml = innerView.Xml;
						}
					}
					catch (XmlException xEx) 
					{
						MessageBox.Show( xEx.Message, XmlTreeView.MessageBoxTitle );
					}
					finally
					{
						childrenXml = string.Empty;
						hitEnd = false;
					}
				}
			}
		}
		#endregion // Properties

		#region Internal methods
		internal void RecurseSubNodes( TreeNode entryNode )
		{
			foreach( TreeNode node in entryNode.Nodes )
			{
				if ( this.Equals( node ) )
				{
					hitStart = true;
				}

				if( !hitEnd && hitStart )
				{
					hitEnd = ( node.Text.EndsWith( "</" + elem.Name + ">" ) && hitStart ) || ( ( (XmlTreeNode) node).ConnectedXmlElement != null && ( (XmlTreeNode) node).ConnectedXmlElement.NodeType == XmlNodeType.Comment );

					childrenXml += indent + node.Text + Environment.NewLine;

					if ( node.Nodes.Count > 0 )
					{
						indent += "    ";
						RecurseSubNodes( node );
					}
				}
			}

			if ( indent.Length > 3 )
			{
				indent = indent.Substring( 0, indent.Length - 4 );
			}
		}
		#endregion // Internal methods
	}
}

File: Contribute/Plugin Samples/VB.NET/XMLParserPlugin/3rd Party Control/XmlTreeNode.vb
Imports System.Xml
Imports System.Windows.Forms
Imports SS.Ynote.Engine.Framework

Namespace XMLTreeView
	''' <summary>
	''' Summary description for XmlTreeNode.
	''' </summary>
	Friend Class XmlTreeNode
		Inherits TreeNode
		#Region "Fields"
		Private elem As XmlNode = Nothing
		Private childrenXml As String = String.Empty
		Private indent As String = String.Empty
		Private hitEnd As Boolean = False
		Private hitStart As Boolean = False
		#End Region

		#Region "Constructors"
		Friend Sub New(text As String, elem As XmlNode)
			MyBase.New(text)
			Me.elem = elem
		End Sub
		#End Region

		#Region "Properties"
		Friend ReadOnly Property ConnectedXmlElement() As XmlNode
			Get
				Return elem
			End Get
		End Property

		Friend ReadOnly Property SelfAndChildren() As String
			Get
				Return childrenXml
			End Get
		End Property

		Public Shadows Property Text() As String
			Get
				Return MyBase.Text
			End Get
			Set
				If value IsNot Nothing AndAlso value.Length > 0 Then
					Try
						If value <> childrenXml Then
							Dim frag As XmlDocumentFragment = elem.OwnerDocument.CreateDocumentFragment()
							frag.InnerXml = value
							elem.ParentNode.ReplaceChild(frag, elem)
							Dim innerView As InnerXmlTreeView = DirectCast(TreeView, InnerXmlTreeView)
							innerView.Xml = innerView.Xml
						End If
					Catch xEx As XmlException
						MessageBox.Show(xEx.Message, XmlTreeView.MessageBoxTitle)
					Finally
						childrenXml = String.Empty
						hitEnd = False
					End Try
				End If
			End Set
		End Property
		#End Region

		#Region "Internal methods"
		Friend Sub RecurseSubNodes(entryNode As TreeNode)
			For Each node As TreeNode In entryNode.Nodes
				If Me.Equals(node) Then
					hitStart = True
				End If

				If Not hitEnd AndAlso hitStart Then
					hitEnd = (node.Text.EndsWith("</" & elem.Name & ">") AndAlso hitStart) OrElse (DirectCast(node, XmlTreeNode).ConnectedXmlElement IsNot Nothing AndAlso DirectCast(node, XmlTreeNode).ConnectedXmlElement.NodeType = XmlNodeType.Comment)

					childrenXml += indent & node.Text & Environment.NewLine

					If node.Nodes.Count > 0 Then
						indent += "    "
						RecurseSubNodes(node)
					End If
				End If
			Next

			If indent.Length > 3 Then
				indent = indent.Substring(0, indent.Length - 4)
			End If
		End Sub
		#End Region
	End Class
End Namespace
